keps/sig-storage/20190122-csi-inline-volumes.md

```diff
@@ -134,27 +132,44 @@ A driver may be able to support either PV/PVC-originated or pod spec originated
+* Persistent inline volume requests will not participate in attach/detach operations that require secrets.
+* Access to pod namespaces are not available during attachment.
+* This means secret references will not be resolved during attachment/detachment.
+* CSI operations during attachment/detachment will receive a nil secret references.
+
+Existing in-tree drivers only fully support attachment/detachment phases with PV/PVC-originated volumes. While adding attachment support for pod-originated is a nice feature, it will only work with the limitations outline above.
+
+### Ephemeral inline mode
+Ephemeral inline volumes are more restricted than their persistent counterparts.  The `CSIDriver.lifecycle = ephemeral` must be set to indicate support for inline ephemeral volumes which will follow the lifecycle of its pod.  For drivers indicating support for ephemeral mode: 
+* Volume requests are assumed to have originated from an inline pod spec.
+* The `volumeHandle` can be omitted. If provided in the pod spec, it will be ignored.
+* The CSI Kubelet plugin will internally generate a `volumeHandle` which is passed to the driver.
 * Ephemeral volumes will only make CSI calls for *mount* and *unmount* volume operations.
 
-#### Omitting inline mode
+During mount/unmount, the CSI Kubelet code will have access to both podUID and pod namespace information. Therefore for
+* Ephemeral inline volume requests will participate in mount/unmount operations.
+* The Kubelet can send mount/unmount rellated calls that require secrets to the driver.
+* Empheral volume requests will have access to both podUID and pod namespaces during mount/unmount operations.
+* Declared secrets references can be fully realized by accessing both secret name and the namespace for the pod where it is used.
+* CSI operations during mount/unmount phases can receive secret values stored in Kubernetes.
+
+### Omitting inline mode setting
 Omitting `CSIDriver.lifecycle` is an indication:
 * The driver can support PV/PVC originated volume (normal operation).
 * If the volume originates from a pod spec, it is assumed that its lifecycle is `persistent`.
 
 ### VolumeHandle generation
-When a driver has its lifecycle set to `ephemeral` (see above), the Kubelet (internal CSI code) will employ a naming strategy to generate the value for the volumeHandle.  The generated value will be a combination of `podUID` and `pod.spec.Volume[x].name` to guarantee uniqueness.
+When a driver has its lifecycle set to `ephemeral` (see above), the Kubelet (internal CSI code) will employ a naming strategy to generate the value for the volumeHandle.  The generated value will be a combination of `podUID` and `pod.spec.Volume[x].name` to guarantee uniqueness.  The generated value will be stable and the Kubelet will be able to regenerate the value as needed during different phases of storage operations.
```

Comment:
unmount never needs to be called outside the pod object being available (like if a pod was force deleted without waiting for unmount)? (or we can recover the pod uid and volume name from on-disk data embedded in file or directory names?)

Reply:
There is mechanism to recover pod information. Currently, the code creates a json file under the pod's directory that is used to reconstruct pod info.